### PR TITLE
MAILBOX-378 Refactor message manager expunge

### PR DESCRIPTION
**Pull request by Arsnael**
As the draft tag says itself, it's still a WIP. I still need to change / add proper tests and maybe refactor some parts of the solution, but you can still have a look if you want.

### CODE REVIEW

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java [hunk 1]
```diff
@@ -117,6 +117,17 @@
...
     Iterator<MessageUid> expunge(MessageRange set, MailboxSession mailboxSession) throws MailboxException;
 
+    /**
+     * Deletes messages
+     *
+     * @param uids
+     *            list of {@link MessageUid} to delete
+     * @param mailboxSession
+     *            not null
+     * @throws MailboxException
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java]: When we write new javadoc we should only add information not only present in the method signature...

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 2]
```diff
@@ -385,6 +408,12 @@ protected MessageMetaData save(Mailbox mailbox, MailboxMessage message) throws M
         return data;
     }
 
+    private List<MessageUid> getUidList(List<MailboxMessage> messages) {
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: Is this method used?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 3]
```diff
@@ -259,17 +259,47 @@ public boolean isModSeqPermanent(MailboxSession session) {
...
+            .metaData(ImmutableSortedMap.copyOf(deletedMessages))
             .build(),
             new MailboxIdRegistrationKey(mailbox.getMailboxId()))
             .block();
-        return uids.keySet().iterator();
+        return deletedMessages.keySet().iterator();
+    }
+
+    private List<MessageUid> retrieveMessagesMarkedForDeletion(final MessageRange messageRange, MailboxSession session) throws MailboxException {
+        final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: final variable spotted

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 4]
```diff
@@ -259,17 +259,47 @@ public boolean isModSeqPermanent(MailboxSession session) {
...
+
+        return messageMapper.execute(
+            () -> messageMapper.retrieveMessagesMarkedForDeletion(getMailboxEntity(), messageRange));
+    }
+
+    @Override
+    public void delete(List<MessageUid> messageUids, MailboxSession mailboxSession) throws MailboxException {
+        Map<MessageUid, MessageMetaData> deletedMessages = deleteMessages(messageUids, mailboxSession);
+
+        eventBus.dispatch(EventFactory.expunged()
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: Isn't that `eventBus dispatch` the exact same compared to the one performed by the expunge?

Can't we extract some common chunks of code?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 5]
```diff
@@ -259,17 +259,47 @@ public boolean isModSeqPermanent(MailboxSession session) {
...
+                .mailboxSession(mailboxSession)
+                .mailbox(getMailboxEntity())
+                .metaData(ImmutableSortedMap.copyOf(deletedMessages))
+                .build(),
+            new MailboxIdRegistrationKey(mailbox.getMailboxId()))
+            .block();
+    }
+
+    private Map<MessageUid, MessageMetaData> deleteMessages(List<MessageUid> messageUids, MailboxSession session) throws MailboxException {
+        final MessageMapper messageMapper = mapperFactory.getMessageMapper(session);
```
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: final

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java [hunk 6]
```diff
@@ -365,7 +365,7 @@ public void findFirstUnseenMessageUidShouldReturnUid2WhenUid2isSeen() throws Mai
         messageMapper.updateFlags(benwaInboxMailbox, new FlagsUpdateCalculator(new Flags(Flags.Flag.SEEN), FlagsUpdateMode.REPLACE), MessageRange.one(message5.getUid()));
         assertThat(messageMapper.findFirstUnseenMessageUid(benwaInboxMailbox)).isEqualTo(message2.getUid());
     }
-
+/*
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: Ouups?

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java [hunk 7]
```diff
@@ -365,7 +365,7 @@ public void findFirstUnseenMessageUidShouldReturnUid2WhenUid2isSeen() throws Mai
         messageMapper.updateFlags(benwaInboxMailbox, new FlagsUpdateCalculator(new Flags(Flags.Flag.SEEN), FlagsUpdateMode.REPLACE), MessageRange.one(message5.getUid()));
```
Comment (chibenwa) [inline mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: We miss tests here for your brand new message manager method

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java (hunk 2 above)
Comment (Arsnael) [reply mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: It is, in `retrieveMessagesMarkedForDeletion`, line 207

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java (hunk 4 above)
Comment (Arsnael) [reply mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: I know, I was just trying to make it work first (thus the draft status for now). That is part of my refactoring list already :)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java (hunk 6 above)
Comment (Arsnael) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: No oupss, the `expungedMarkedForDeletion` method has been removed, so I need to refactor those tests to take the new methods into account, but I wanted to be sure that it was compiling and that I didn't break anything else at that stage... :)

File: mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java (hunk 7 above)
Comment (Arsnael) [reply mailbox/store/src/test/java/org/apache/james/mailbox/store/mail/model/MessageMapperTest.java]: I know, will work on the tests today

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java [hunk 8]
```diff
@@ -117,6 +117,17 @@
      */
     Iterator<MessageUid> expunge(MessageRange set, MailboxSession mailboxSession) throws MailboxException;
 
+    /**
+     * Deletes messages
```
Comment (trantienduchn) [inline mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java]: maybe you can explain a bit what is the different between delete and expunge method,.

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 9]
```diff
@@ -197,42 +198,64 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
+            MessageUid to = range.getUidTo();
+            List<MailboxMessage> messages = findDeletedMessages(range, from, to, mailboxId);
+            data.putAll(createMetaData(messages));
+            deleteDeletedMessages(range, from, to, mailboxId);
+        });
+
+        return data;
+    }
+
+    private List<MailboxMessage> findDeletedMessages(MessageRange messageRange, MessageUid from, MessageUid to, JPAId mailboxId) {
```
Comment (trantienduchn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: I think this method only need one parameter `MessageRange`, the others `from` and `to` can be easily retrieved from the range.

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 10]
```diff
@@ -197,42 +198,64 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
+            case FROM:
+                return findDeletedMessagesInMailboxAfterUID(mailboxId, from);
+            case ALL:
+            default:
+                return findDeletedMessagesInMailbox(mailboxId);
+        }
+    }
 
-            switch (set.getType()) {
+    private void deleteDeletedMessages(MessageRange messageRange, MessageUid from, MessageUid to, JPAId mailboxId) {
```
Comment (trantienduchn) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: idem

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java [hunk 11]
```diff
@@ -207,39 +209,60 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
-            break;
-        case ONE:
-            results = findDeletedMessageInMailboxWithUID(mailbox, from);
-            break;
-        case RANGE:
-            results = findMessagesInMailboxBetweenUIDs(mailbox, MaildirMessageName.FILTER_DELETED_MESSAGES, from, to,
-                    -1);
-            break;
+    public List<MessageUid> retrieveMessagesMarkedForDeletion(Mailbox mailbox, MessageRange messageRange) throws MailboxException {
+        List<MessageUid> results = new ArrayList<>();
```
Comment (trantienduchn) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java]: looks like you dont need this variable.

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java [hunk 12]
```diff
@@ -207,39 +209,60 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
+        MessageUid from = messageRange.getUidFrom();
+        MessageUid to = messageRange.getUidTo();
+        List<MailboxMessage> messages = findDeletedMessages(mailbox, messageRange, from, to);
```
Comment (trantienduchn) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java]: tiny comment: if `from`, `to` are not used in anywhere else, you could inline them to this method call.

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java [hunk 13]
```diff
@@ -207,39 +209,60 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
+            MessageUid from = range.getUidFrom();
+            MessageUid to = range.getUidTo();
+            List<MailboxMessage> messages = findDeletedMessages(mailbox, range, from, to);
+            data.putAll(deleteDeletedMessages(mailbox, messages));
         }
+
+        return data;
+    }
+
+    private List<MailboxMessage> findDeletedMessages(Mailbox mailbox, MessageRange messageRange, MessageUid from, MessageUid to) throws MailboxException {
```
Comment (trantienduchn) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java]: idem for method parameters simplification

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageMapper.java [hunk 14]
```diff
@@ -136,24 +134,32 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
         return filteredResult;
     }
 
     @Override
-    public Flags getApplicableFlag(Mailbox mailbox) throws MailboxException {
+    public Map<MessageUid, MessageMetaData> deleteMessages(Mailbox mailbox, List<MessageUid> uids) {
+        return getMembershipByUidForMailbox(mailbox).values()
+            .stream()
+            .filter(message -> uids.contains(message.getUid()))
+            .sorted()
```
Comment (trantienduchn) [inline mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageMapper.java]: why did you need `sorted()`?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java [hunk 15]
```diff
@@ -259,17 +259,47 @@ public boolean isModSeqPermanent(MailboxSession session) {
...
-            .metaData(ImmutableSortedMap.copyOf(uids))
+            .metaData(ImmutableSortedMap.copyOf(deletedMessages))
             .build(),
             new MailboxIdRegistrationKey(mailbox.getMailboxId()))
             .block();
-        return uids.keySet().iterator();
+        return deletedMessages.keySet().iterator();
+    }
+
+    private List<MessageUid> retrieveMessagesMarkedForDeletion(final MessageRange messageRange, MailboxSession session) throws MailboxException {
```
Comment (trantienduchn) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreMessageManager.java]: is final here needed?

File: server/protocols/protocols-pop3/src/main/java/org/apache/james/pop3server/mailbox/MailboxAdapter.java [hunk 16]
```diff
@@ -174,15 +172,11 @@ public void remove(String... uids) throws IOException {
...
         try {
             mailboxManager.startProcessingRequest(session);
-            for (MessageRange range : ranges) {
-                manager.setFlags(new Flags(Flags.Flag.DELETED), MessageManager.FlagsUpdateMode.ADD, range, session);
-                manager.expunge(range, session);
-            }
+            manager.delete(uidList, session);
         } catch (MailboxException e) {
-            throw new IOException("Unable to remove messages for ranges " + ranges);
+            throw new IOException("Unable to remove messages", e);
```
Comment (trantienduchn) [inline server/protocols/protocols-pop3/src/main/java/org/apache/james/pop3server/mailbox/MailboxAdapter.java]: I think passing the details of the range is better than passing no information to the log.

Comment (Arsnael) [reply server/protocols/protocols-pop3/src/main/java/org/apache/james/pop3server/mailbox/MailboxAdapter.java]: Except there is no range anymore here, just a list of uids?

File: mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java (hunk 8 above)
Comment (chibenwa) [reply mailbox/api/src/main/java/org/apache/james/mailbox/MessageManager.java]: :+1:  Maybe we should cover the side effects (events ... )

File: server/protocols/protocols-pop3/src/main/java/org/apache/james/pop3server/mailbox/MailboxAdapter.java (hunk 16 above)
Comment (chibenwa) [reply server/protocols/protocols-pop3/src/main/java/org/apache/james/pop3server/mailbox/MailboxAdapter.java]: Can we rather pass an Immutable list of uid? This information is important...

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java (hunk 11 above)
Comment (Arsnael) [reply mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java]: Something I missed after refactoring... Well seen :)

File: mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageMapper.java (hunk 14 above)
Comment (Arsnael) [reply mailbox/memory/src/main/java/org/apache/james/mailbox/inmemory/mail/InMemoryMessageMapper.java]: No need, you are right

File: mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java [hunk 17]
```diff
@@ -197,42 +198,66 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
+
+        switch (messageRange.getType()) {
+            case ONE:
+                return findDeletedMessagesInMailboxWithUID(mailboxId, from);
+            case RANGE:
+                return findDeletedMessagesInMailboxBetweenUIDs(mailboxId, from, to);
+            case FROM:
+                return findDeletedMessagesInMailboxAfterUID(mailboxId, from);
+            case ALL:
+            default:
```
Comment (chibenwa) [inline mailbox/jpa/src/main/java/org/apache/james/mailbox/jpa/mail/JPAMessageMapper.java]: If the type of range is onknown it is safer to throw a runtime exception so that the programmer get that something is wrong rather than trying to enforce wrong behaviours....

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java [hunk 18]
```diff
@@ -207,39 +209,58 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
...
+    }
+
+    private Map<MessageUid, MessageMetaData> deleteDeletedMessages(Mailbox mailbox, List<MailboxMessage> messages) throws MailboxException {
         Map<MessageUid, MessageMetaData> uids = new HashMap<>();
-        for (MailboxMessage m : results) {
-            MessageUid uid = m.getUid();
-            uids.put(uid, m.metaData());
-            delete(mailbox, m);
+
+        for (MailboxMessage message : messages) {
```
Comment (chibenwa) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java]: Smells like this can be written with a stream <3

**Comment by Arsnael** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

File: mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java [hunk 19]
```diff
@@ -239,25 +240,21 @@ public MessageUid findFirstUnseenMessageUid(Mailbox mailbox) throws MailboxExcep
             case FROM:
                 return findMessagesInMailboxBetweenUIDs(mailbox, MaildirMessageName.FILTER_DELETED_MESSAGES, from, null, -1);
             case ALL:
-            default:
                 return findMessagesInMailbox(mailbox, MaildirMessageName.FILTER_DELETED_MESSAGES, -1);
+            default:
+                throw new RuntimeException();
```
Comment (rouazana) [inline mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java]: can you add at least a description?

Comment (Arsnael) [reply mailbox/maildir/src/main/java/org/apache/james/mailbox/maildir/mail/MaildirMessageMapper.java]: https://github.com/linagora/james-project/pull/2189
